cpp/cuproj/include/cuproj/operation/transverse_mercator.cuh

```diff
@@ -0,0 +1,518 @@
+ copy of this software and associated documentation files (the "Software"),
+ to deal in the Software without restriction, including without limitation
+ the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ and/or sell copies of the Software, and to permit persons to whom the
+ Software is furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included
+ in all copies or substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ DEALINGS IN THE SOFTWARE.
+*/
+
+// The following text is taken from the PROJ file
+// [tmerc.cpp](https://github.com/OSGeo/PROJ/blob/9.2/src/projections/tmerc.cpp)
+// see the file LICENSE_PROJ in the cuspatial repository root directory for the original PROJ
+// license text.
+/*****************************************************************************/
+//
+//                  Exact Transverse Mercator functions
+//
+//
+// The code in this file is largly based upon procedures:
+//
+// Written by: Knud Poder and Karsten Engsager
+//
+// Based on math from: R.Koenig and K.H. Weise, "Mathematische
+// Grundlagen der hoeheren Geodaesie und Kartographie,
+// Springer-Verlag, Berlin/Goettingen" Heidelberg, 1951.
+//
+// Modified and used here by permission of Reference Networks
+// Division, Kort og Matrikelstyrelsen (KMS), Copenhagen, Denmark
+//
+/*****************************************************************************/
+
+#pragma once
+
+#include <cuproj/ellipsoid.hpp>
+#include <cuproj/operation/operation.cuh>
+#include <cuproj/projection_parameters.hpp>
+
+#include <thrust/iterator/transform_iterator.h>
+
+#include <assert.h>
+
+namespace cuproj {
+
+namespace detail {
+
+/**
+ * @brief Evaluate Gaussian<-->Geographic trigonometric series
+ *
+ * @tparam T data type
+ * @param p1  pointer to the first element of the array
+ * @param len_p1  length of the array
+ * @param B The argument to the trigonometric series
+ * @param cos_2B precomputed cos(2*B)
+ * @param sin_2B precomputed sin(2*B)
+ * @return The value of the trigonometric series at B
+ */
+template <typename T>
+inline static __host__ __device__ T gatg(T const* p1, int len_p1, T B, T cos_2B, T sin_2B)
+{
+  T h = 0, h1, h2 = 0;
+
+  T const two_cos_2B = 2 * cos_2B;
+  T const* p         = p1 + len_p1;
+  h1                 = *--p;
+  while (p - p1) {
+    h  = -h2 + two_cos_2B * h1 + *--p;
+    h2 = h1;
+    h1 = h;
+  }
+  return (B + h * sin_2B);
+}
+
+// Complex Clenshaw summation
```

Comment:
```suggestion
```

Reply:
Do these comments serve a purpose? Otherwise they look redundant to `@brief`.

Reply:
Removed